File: amqpAPI/java/impl/src/uk/ac/soton/itinnovation/experimedia/arch/ecc/amqpAPI/impl/amqp/AMQPBasicSubscriptionProcessor.java
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2012
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          11-Jul-2012
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

package uk.ac.soton.itinnovation.experimedia.arch.ecc.amqpAPI.impl.amqp;

import com.rabbitmq.client.*;
import java.io.IOException;
import org.apache.log4j.Logger;




public class AMQPBasicSubscriptionProcessor extends DefaultConsumer
{
  private final Logger subProcLogger = Logger.getLogger( AMQPBasicSubscriptionProcessor.class );
  
  private String              queueName;
  private AMQPMessageDispatch messageDispatch;

  
  public AMQPBasicSubscriptionProcessor( Channel channel,
                                         String qName,
                                         AMQPMessageDispatch dispatch )
  {
    super( channel );
    
    queueName       = qName;
    messageDispatch = dispatch;
  }
  
  @Override
  public void handleDelivery( String consumerTag,
                              Envelope envelope,
                              AMQP.BasicProperties properties,
                              byte[] body )
  {
    messageDispatch.addMessage( queueName, body );
    
    try { getChannel().basicAck( envelope.getDeliveryTag(), true ); }
    catch (IOException ioe)
    { subProcLogger.error( "Could not send AMQP acknowledgement" ); }
  }
}

File: extensions/cppClientAPI/amqpAPI/amqp-Impl/amqp/AMQPBasicSubscriptionProcessor.cpp
/////////////////////////////////////////////////////////////////////////
//
// © University of Southampton IT Innovation Centre, 2013
//
// Copyright in this software belongs to University of Southampton
// IT Innovation Centre of Gamma House, Enterprise Road, 
// Chilworth Science Park, Southampton, SO16 7NS, UK.
//
// This software may not be used, sold, licensed, transferred, copied
// or reproduced in whole or in part in any manner or form or in or
// on any media by any person other than in accordance with the terms
// of the Licence Agreement supplied with the software, or otherwise
// without the prior written consent of the copyright owners.
//
// This software is distributed WITHOUT ANY WARRANTY, without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE, except where stated in the Licence Agreement supplied with
// the software.
//
//      Created By :            Simon Crowle
//      Created Date :          15-May-2013
//      Created for Project :   EXPERIMEDIA
//
/////////////////////////////////////////////////////////////////////////

#include "stdafx.h"

#include "AMQPBasicSubscriptionProcessor.h"

#include "ECCUtils.h"

#include <amqp.h>

using namespace AmqpClient;

using namespace std;
using namespace boost;
using namespace boost::uuids;





namespace ecc_amqpAPI_impl
{
  
  AMQPBasicSubscriptionProcessor::AMQPBasicSubscriptionProcessor( AMQPBasicSubscriptionService::ptr_t service,
                                                                  Channel::ptr_t                      channel,
                                                                  const String&                       qName,
                                                                  AMQPMessageDispatch::ptr_t          dispatch )
  : processorID( createRandomUUID() )
  {
    subscriptionService = service;
    amqpChannelImpl     = channel;
    queueName           = qName;
    messageDispatch     = dispatch;
  }

  AMQPBasicSubscriptionProcessor::~AMQPBasicSubscriptionProcessor()
  {
    if ( subscriptionService )
      subscriptionService->unsubscribe( processorID );
  }

  void AMQPBasicSubscriptionProcessor::initialiseSubscription()
  {
    if ( subscriptionService )
      subscriptionService->subscribe( shared_from_this() );
  }

  uuids::uuid AMQPBasicSubscriptionProcessor::getProcessorID()
  { return processorID; }

  void AMQPBasicSubscriptionProcessor::handleBasicMessage( Envelope::ptr_t envolope )
  {
    if ( envolope )
    {
      BasicMessage::ptr_t msg = envolope->Message();

      if ( msg )
      {
        const string payload = msg->Body();
        
        if ( !payload.empty() )
          messageDispatch->addMessage( toNarrow(queueName), payload );
      }
      
      amqpChannelImpl->BasicAck( envolope );
    }
  }

}